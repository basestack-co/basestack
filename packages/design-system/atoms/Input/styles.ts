import styled, { css } from "styled-components";
import { rem } from "polished";
import { space, compose, layout } from "styled-system";
import Icon from "../Icon";

export const Container = styled.div`
  ${compose(space, layout)};
  position: relative;
  display: flex;
  align-items: center;
`;

export const StyledInput = styled.input<{
  isDarker: boolean;
  format?: "small" | "normal";
  hasLeftIcon: boolean;
  hasRightIcon: boolean;
  hasError: boolean;
  disabled: boolean;
}>`
  border: none;
  background-color: ${({ theme, isDarker, hasError }) =>
    hasError
      ? theme.colors.red50
      : isDarker
      ? theme.colors.gray100
      : theme.colors.gray50};
  height: ${({ format }) => (format === "normal" ? rem("44px") : rem("36px"))};
  border-radius: ${rem("4px")};
  font-size: ${rem("14px")};
  font-weight: 400;
  width: 100%;
  padding: 0 ${rem("16px")};
  color: ${({ theme }) => theme.colors.black};

  &::placeholder {
    color: ${({ theme }) => theme.colors.gray500};
  }

  &:focus {
    outline: 2px solid
      ${({ theme, hasError }) =>
        hasError ? theme.colors.red200 : theme.colors.black};
  }

  ${({ hasLeftIcon }) =>
    hasLeftIcon &&
    css`
      padding-left: ${rem("44px")};
    `};

  ${({ hasRightIcon }) =>
    hasRightIcon &&
    css`
      padding-right: ${rem("44px")};
    `};

  ${({ disabled }) =>
    disabled &&
    css`
      opacity: 0.6;
    `};
`;

export const LeftIcon = styled(Icon)`
  position: absolute;
  left: ${rem("12px")};
`;

export const RightIcon = styled(Icon)`
  position: absolute;
  right: ${rem("12px")};
`;
