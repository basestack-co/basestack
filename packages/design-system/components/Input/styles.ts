import styled, { css } from "styled-components";
import { rem } from "polished";
import { space, compose, layout } from "styled-system";

export const Container = styled.div.withConfig({
  shouldForwardProp: (prop) => prop !== "maxWidth",
})`
  ${compose(space, layout)};
  position: relative;
  display: flex;
  align-items: center;
`;

export const StyledInput = styled.input.withConfig({
  shouldForwardProp: (prop) =>
    ![
      "hasError",
      "isDarker",
      "iconPlacement",
      "hasRightIcon",
      "hasLeftIcon",
    ].includes(prop),
})<{
  isDarker: boolean;
  format?: "small" | "normal";
  hasLeftIcon: boolean;
  hasRightIcon: boolean;
  hasError: boolean;
  disabled: boolean;
}>`
  border: none;
  background-color: ${({ theme, isDarker, hasError }) =>
    hasError
      ? theme.input.error.backgroundColor
      : isDarker
        ? theme.input.isDarker.backgroundColor
        : theme.input.backgroundColor};
  height: ${({ format }) => (format === "normal" ? rem("44px") : rem("36px"))};
  border-radius: ${rem("4px")};
  font-size: ${rem("14px")};
  font-weight: 400;
  width: 100%;
  padding: 0 ${rem("16px")};
  color: ${({ theme }) => theme.input.color};

  &::placeholder {
    color: ${({ theme }) => theme.input.placeholder.color};
  }

  &:focus {
    outline: 2px solid
      ${({ theme, hasError }) =>
        hasError ? theme.input.error.focus.outline : theme.input.focus.outline};
  }

  ${({ hasLeftIcon }) =>
    hasLeftIcon &&
    css`
      padding-left: ${rem("44px")};
    `};

  ${({ hasRightIcon }) =>
    hasRightIcon &&
    css`
      padding-right: ${rem("44px")};
    `};

  ${({ disabled }) =>
    disabled &&
    css`
      opacity: 0.6;
    `};
`;

export const IconContainer = styled.div.withConfig({
  shouldForwardProp: (prop) => prop !== "iconPlacement",
})<{ iconPlacement?: "left" | "right" }>`
  display: inline-flex;
  position: absolute;
  left: ${({ iconPlacement }) =>
    iconPlacement === "left" ? rem("12px") : "initial"};
  right: ${({ iconPlacement }) =>
    iconPlacement === "right" ? rem("12px") : "initial"};
`;
