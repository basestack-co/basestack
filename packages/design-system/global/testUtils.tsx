import { type RenderOptions, render } from "@testing-library/react";
import type React from "react";
// Styles
import { ThemeProvider } from "styled-components";
import theme from "../theme/lightTheme";

interface Props {
  children: React.ReactNode;
}

const WithThemeProvider = ({ children }: Props) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
);

const renderWithTheme = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">,
) => render(ui, { wrapper: WithThemeProvider, ...options });

export { renderWithTheme };
