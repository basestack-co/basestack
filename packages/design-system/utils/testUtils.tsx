import defaultTheme from "@basestack/design-system/theme/lightTheme";
import isPropValid from "@emotion/is-prop-valid";
import { RenderResult, render } from "@testing-library/react";
import React from "react";
import { StyleSheetManager, ThemeProvider } from "styled-components";

const Provider = ({
  children,
  theme = defaultTheme,
}: {
  children: React.ReactNode;
  theme: any;
}) => (
  <ThemeProvider theme={theme}>
    <StyleSheetManager shouldForwardProp={isPropValid}>
      {children}
    </StyleSheetManager>
  </ThemeProvider>
);

export const renderWithTheme = (
  ui: React.ReactElement,
  options?: any,
): RenderResult => render(ui, { wrapper: Provider, ...options });
