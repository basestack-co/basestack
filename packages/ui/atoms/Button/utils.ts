import { Size, Target, Type, Variant } from "@/components/atoms/Button/types";
import { css, DefaultTheme } from "styled-components";
import { darken, rem, transparentize } from "polished";

export const handleButtonSize = (size?: Size) => {
  switch (size) {
    case "large":
      return css`
        height: ${rem("60px")};
        font-size: ${rem("18px")};
        line-height: ${rem("25px")};
        padding: 0 ${rem("40px")};
      `;
    case "medium":
      return css`
        height: ${rem("54px")};
        font-size: ${rem("16px")};
        line-height: ${rem("22px")};
        padding: 0 ${rem("34px")};
      `;
    case "normal":
      return css`
        height: ${rem("48px")};
        font-size: ${rem("16px")};
        line-height: ${rem("22px")};
        padding: 0 ${rem("30px")};
      `;
    default:
    case "small":
      return css`
        height: ${rem("46px")};
        font-size: ${rem("14px")};
        line-height: ${rem("19px")};
        padding: 0 ${rem("24px")};
      `;
    case "xSmall":
      return css`
        height: ${rem("42px")};
        font-size: ${rem("14px")};
        line-height: ${rem("19px")};
        padding: 0 ${rem("20px")};
      `;
  }
};

export const handleButtonsColors = (theme: DefaultTheme) => ({
  primary: {
    outline: {
      color: theme.colors.primary,
      borderColor: theme.colors.primary,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.primary,
    },
  },
  secondary: {
    outline: {
      color: theme.colors.secondary,
      borderColor: theme.colors.secondary,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.secondary,
    },
  },
  tertiary: {
    outline: {
      color: theme.colors.gray3,
      borderColor: theme.colors.gray3,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.primary,
      borderColor: "transparent",
      backgroundColor: theme.colors.gray6,
    },
  },
  neutral: {
    outline: {
      color: theme.colors.gray2,
      borderColor: theme.colors.gray2,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.primary,
      borderColor: "transparent",
      backgroundColor: "transparent",
    },
  },
  success: {
    outline: {
      color: theme.colors.success,
      borderColor: theme.colors.success,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.success,
    },
  },
  warning: {
    outline: {
      color: theme.colors.warning,
      borderColor: theme.colors.warning,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.warning,
    },
  },
  danger: {
    outline: {
      color: theme.colors.error,
      borderColor: theme.colors.error,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.error,
    },
  },
});

export const darkenColors = (
  theme: DefaultTheme,
  amount: number,
  type: Type,
  variant: Variant,
  target: Target
) => {
  return darken(
    amount || 0.1,
    handleButtonsColors(theme)[type || "primary"][variant || "filled"][target]
  );
};

export const lightenColors = (
  theme: DefaultTheme,
  amount: number,
  type: Type,
  variant: Variant,
  target: Target
) => {
  return transparentize(
    amount || 0.6,
    handleButtonsColors(theme)[type || "primary"][variant || "filled"][target]
  );
};
