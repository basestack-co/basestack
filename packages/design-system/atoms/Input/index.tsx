import React, { memo } from "react";
import { useTheme } from "styled-components";
import { SpaceProps, LayoutProps } from "styled-system";
import { Container, StyledInput, LeftIcon, RightIcon } from "./styles";

export interface InputProps extends SpaceProps, LayoutProps {
  /**
   * Optional Icon
   */
  icon?: string;
  /**
   * Input type
   */
  type?: string;
  /**
   * Set input to darker
   */
  isDarker?: boolean;
  /**
   * Input size
   */
  size?: "small" | "normal";
  /**
   * Icon Placement
   */
  iconPlacement?: "left" | "right";
  /**
   * Input placeholder
   */
  placeholder: string;
  /**
   * Input onChange
   */
  onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  /**
   * Input optional testID
   */
  testId?: string;
}

const Input: React.FC<InputProps> = ({
  icon,
  iconPlacement = "left",
  type = "text",
  placeholder,
  isDarker = false,
  size = "normal",
  testId = "input-container",
  onChange,
  ...props
}) => {
  const theme = useTheme();
  const hasLeftIcon = !!icon && iconPlacement === "left";
  const hasRightIcon = !!icon && iconPlacement === "right";

  return (
    <Container data-testid={testId} {...props}>
      {hasLeftIcon && (
        <LeftIcon icon={icon} size="medium" color={theme.colors.gray500} />
      )}
      <StyledInput
        data-test-id="input"
        onChange={onChange}
        placeholder={placeholder}
        type={type}
        isDarker={isDarker}
        size={size}
        hasLeftIcon={hasLeftIcon}
        hasRightIcon={hasRightIcon}
      />
      {hasRightIcon && (
        <RightIcon icon={icon} size="medium" color={theme.colors.gray500} />
      )}
    </Container>
  );
};

export default memo(Input);
