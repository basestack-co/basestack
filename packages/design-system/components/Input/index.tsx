import React, { memo } from "react";
import { useTheme } from "styled-components";
import { SpaceProps, LayoutProps } from "styled-system";
import Icon from "../Icon";
import { Container, StyledInput, IconContainer } from "./styles";

export interface InputProps extends SpaceProps, LayoutProps {
  /**
   * Optional Icon
   */
  icon?: string;
  /**
   * Input type
   */
  type?: string;
  /**
   * Set input to darker
   */
  isDarker?: boolean;
  /**
   * Input size
   */
  format?: "small" | "normal";
  /**
   * Icon Placement
   */
  iconPlacement?: "left" | "right";
  /**
   * Input placeholder
   */
  placeholder: string;
  /**
   * Input value
   */
  value: string;
  /**
   * Input name
   */
  name: string;
  /**
   * Input onChange
   */
  onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  /**
   * Input onBlur
   */
  onBlur?: (event: React.FocusEvent<HTMLInputElement>) => void;
  /**
   * Input onFocus
   */
  onFocus?: (event: React.FocusEvent<HTMLInputElement>) => void;
  /**
   * Input optional testID
   */
  testId?: string;
  /**
   * Input error styles
   */
  hasError?: boolean;
  /**
   * Input disabled state
   */
  isDisabled?: boolean;
  /**
   * Input autoComplete on or off
   */
  autoComplete?: string;
  /**
   * Input min chars
   */
  min?: string;
  /**
   * Input max chars
   */
  max?: string;
}

const Input: React.FC<InputProps> = ({
  icon,
  iconPlacement = "left",
  type = "text",
  placeholder,
  name,
  value,
  isDarker = false,
  format = "normal",
  testId = "input-container",
  onChange,
  onBlur,
  onFocus,
  hasError = false,
  isDisabled = false,
  autoComplete,
  min,
  max,
  ...props
}) => {
  const theme = useTheme();
  const hasLeftIcon = !!icon && iconPlacement === "left";
  const hasRightIcon = !!icon && iconPlacement === "right";

  const iconColor = hasError
    ? theme.input.error.icon.color
    : theme.input.icon.color;

  return (
    <Container data-testid={testId} {...props}>
      {hasLeftIcon && (
        <IconContainer iconPlacement={iconPlacement}>
          <Icon icon={icon} size="medium" color={iconColor} />
        </IconContainer>
      )}
      <StyledInput
        data-testid="input"
        onChange={onChange}
        onBlur={onBlur}
        onFocus={onFocus}
        placeholder={placeholder}
        name={name}
        value={value}
        type={type}
        isDarker={isDarker}
        format={format}
        hasLeftIcon={hasLeftIcon}
        hasRightIcon={hasRightIcon}
        hasError={hasError}
        disabled={isDisabled}
        autoComplete={autoComplete}
        min={min}
        max={max}
      />
      {hasRightIcon && (
        <IconContainer iconPlacement={iconPlacement}>
          <Icon icon={icon} size="medium" color={iconColor} />
        </IconContainer>
      )}
    </Container>
  );
};

export default memo(Input);
