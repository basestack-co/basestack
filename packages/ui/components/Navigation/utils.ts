// Types
import { Translate } from "next-translate";
import { NextRouter } from "next/router";
// Utils
import { config } from "@basestack/utils";
import { signOut } from "next-auth/react";

export const getInternalLinks = (t: Translate, formId: string) => {
  return [
    {
      icon: "flag",
      activeText: "Submissions",
      to: `/form/${formId}/submissions`,
      text: t("internal.submissions"),
    },
    {
      icon: "flag",
      activeText: "Setup",
      to: `/form/${formId}/setup`,
      text: t("internal.setup"),
    },
    {
      icon: "settings",
      activeText: "Settings",
      to: `/form/${formId}/settings/general`,
      text: t("internal.settings"),
    },
  ];
};

export const getExternalLinks = (t: Translate) => {
  return [
    {
      icon: "description",
      activeText: "Documentation",
      to: config.urls.docs.flags.base,
      isExternal: true,
      text: t("external.docs"),
    },
    {
      icon: "link",
      activeText: "Github",
      to: config.urls.repo,
      isExternal: true,
      text: t("external.github"),
    },
  ];
};

export const getAvatarDropdownList = (
  t: Translate,
  router: NextRouter,
  onCreateForm: () => void,
) => {
  return [
    {
      icon: "add_circle",
      text: t("create.form"),
      onClick: onCreateForm,
      separator: true,
    },
    {
      icon: "settings",
      text: t("dropdown.settings"),
      onClick: () =>
        router.push({
          pathname: "/user/profile/settings",
        }),
    },
    {
      icon: "logout",
      text: t("dropdown.logout"),
      onClick: signOut,
    },
  ];
};
