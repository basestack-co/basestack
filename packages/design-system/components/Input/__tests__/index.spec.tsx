import { rem } from "polished";
import { cleanup } from "@testing-library/react";
import { renderWithTheme } from "../../../utils/testUtils";
import Input from "..";

describe("Input Atom tests", () => {
  afterEach(cleanup);

  test("should render Input correctly", () => {
    const { asFragment } = renderWithTheme(
      <Input
        placeholder="Search here..."
        onChange={jest.fn()}
        name="feature"
        value=""
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("should render input container", () => {
    const { getByTestId } = renderWithTheme(
      <Input
        placeholder="Search here..."
        onChange={jest.fn()}
        name="feature"
        value=""
      />,
    );
    const container = getByTestId("input-container");

    expect(container).toHaveStyle(`position: relative`);
    expect(container).toHaveStyle(`display: flex`);
    expect(container).toHaveStyle(`align-items: center`);
  });

  test("should render normal input", () => {
    const { getByTestId } = renderWithTheme(
      <Input
        placeholder="Search here..."
        onChange={jest.fn()}
        format="normal"
        name="feature"
        value=""
      />,
    );
    const input = getByTestId("input");

    expect(input).toHaveStyle(`background-color: #F6F6F6`);
    expect(input).toHaveStyle(`height: ${rem("44px")}`);
    expect(input).toHaveStyle(`font-size: ${rem("14px")}`);
    expect(input).toHaveStyle(`width: 100%`);
    expect(input).toHaveStyle(`padding: 0 ${rem("16px")}`);
  });

  test("should render small input", () => {
    const { getByTestId } = renderWithTheme(
      <Input
        format="small"
        placeholder="Search here..."
        onChange={jest.fn()}
        name="feature"
        value=""
      />,
    );
    const input = getByTestId("input");

    expect(input).toHaveStyle(`height: ${rem("36px")}`);
  });

  test("should render input with left icon", () => {
    const { getByTestId } = renderWithTheme(
      <Input
        placeholder="Search here..."
        icon="search"
        iconPlacement="left"
        onChange={jest.fn()}
        name="feature"
        value=""
      />,
    );
    const input = getByTestId("input");
    const iconContainer = getByTestId("icon-container");
    const icon = getByTestId("icon");

    expect(input).toHaveStyle(`padding-left: ${rem("44px")}`);
    expect(iconContainer).toHaveStyle(`position: absolute`);
    expect(iconContainer).toHaveStyle(`left:  ${rem("12px")}`);
    expect(icon).toBeVisible();
  });

  test("should render input with right icon", () => {
    const { getByTestId } = renderWithTheme(
      <Input
        placeholder="Search here..."
        icon="search"
        iconPlacement="right"
        onChange={jest.fn()}
        name="feature"
        value=""
      />,
    );
    const input = getByTestId("input");
    const iconContainer = getByTestId("icon-container");
    const icon = getByTestId("icon");

    expect(input).toHaveStyle(`padding-right: ${rem("44px")}`);
    expect(iconContainer).toHaveStyle(`position: absolute`);
    expect(iconContainer).toHaveStyle(`right:  ${rem("12px")}`);
    expect(icon).toBeVisible();
  });
});
