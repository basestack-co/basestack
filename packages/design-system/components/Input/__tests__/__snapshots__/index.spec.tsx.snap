// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input tests should render Input correctly 1`] = `
<DocumentFragment>
  .c0 {
  position: relative;
  display: flex;
  align-items: center;
}

.c1 {
  border: none;
  background-color: #F6F6F6;
  height: 2.75rem;
  border-radius: 0.25rem;
  font-size: 0.875rem;
  font-weight: 400;
  width: 100%;
  padding: 0 1rem;
  color: #000000;
}

.c1::placeholder {
  color: #757575;
}

.c1:focus {
  outline: 2px solid #000000;
}

<div
    class="c0"
    data-testid="input-container"
  >
    <input
      class="c1"
      data-testid="input"
      format="normal"
      name="feature"
      placeholder="Search here..."
      type="text"
      value=""
    />
  </div>
</DocumentFragment>
`;
